yt_dlp/downloader/websocket.py

```diff
@@ -1,53 +1,106 @@
 from ..compat import asyncio
 from ..dependencies import websockets
 
 
-class FFmpegSinkFD(FileDownloader):
+class AsyncSinkFD(FileDownloader):
+    async def connect(self, stdin, info_dict):
+        try:
+            await self.real_connection(stdin, info_dict)
+        except OSError:
+            pass
+        finally:
+            with contextlib.suppress(OSError):
+                stdin.flush()
+                stdin.close()
+
+    async def real_connection(self, sink, info_dict):
+        """ Override this in subclasses """
+        raise NotImplementedError('This method must be implemented by subclasses')
+
+
+class FFmpegSinkFD(AsyncSinkFD):
     """ A sink to ffmpeg for downloading fragments in any form """
 
     def real_download(self, filename, info_dict):
         info_copy = info_dict.copy()
         info_copy['url'] = '-'
-
-        async def call_conn(proc, stdin):
-            try:
-                await self.real_connection(stdin, info_dict)
-            except OSError:
-                pass
-            finally:
-                with contextlib.suppress(OSError):
-                    stdin.flush()
-                    stdin.close()
-                os.kill(os.getpid(), signal.SIGINT)
+        connect = self.connect
 
         class FFmpegStdinFD(FFmpegFD):
             @classmethod
             def get_basename(cls):
                 return FFmpegFD.get_basename()
 
             def on_process_started(self, proc, stdin):
-                thread = threading.Thread(target=asyncio.run, daemon=True, args=(call_conn(proc, stdin), ))
+                thread = threading.Thread(target=asyncio.run, daemon=True, args=(connect(stdin, info_dict), ))
                 thread.start()
 
         return FFmpegStdinFD(self.ydl, self.params or {}).download(filename, info_copy)
 
-    async def real_connection(self, sink, info_dict):
-        """ Override this in subclasses """
-        raise NotImplementedError('This method must be implemented by subclasses')
 
+class FileSinkFD(AsyncSinkFD):
+    """ A sink to a file for downloading fragments in any form """
+    def real_download(self, filename, info_dict):
+        tempname = self.temp_name(filename)
+        try:
+            with open(tempname, 'wb') as w:
+                started = time.time()
+                status = {
+                    'filename': info_dict.get('_filename'),
+                    'status': 'downloading',
+                    'elapsed': 0,
+                    'downloaded_bytes': 0,
+                }
+                self._hook_progress(status, info_dict)
+
+                thread = threading.Thread(target=asyncio.run, daemon=True, args=(self.connect(w, info_dict), ))
+                thread.start()
+                time_and_size, avg_len = [], 10
+                while thread.is_alive():
+                    time.sleep(0.1)
+
+                    downloaded, curr = w.tell(), time.time()
+                    # taken from ffmpeg attachment
```

Comment:
?

Reply:
To show progress, it is taken from https://github.com/ytdl-patched/ytdl-patched/blob/ytdlp/yt_dlp/postprocessor/_attachments.py